sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -39,81 +39,161 @@
-        responses.put(response)
-
-      thread_pool.submit(do_instruction, work_request).add_done_callback(
-          functools.partial(handle_response, work_request))
-
-    logging.info("No more requests from control plane")
-    logging.info("SDK Harness waiting for in-flight requests to complete")
+      # Name spacing the request method with 'request_'. The called method
+      # will be like self.request_register(request)
+      getattr(self, 'request_' + request_type)(work_request)
+
+    logging.info('No more requests from control plane')
+    logging.info('SDK Harness waiting for in-flight requests to complete')
     # Wait until existing requests are processed.
     self._progress_thread_pool.shutdown()
-    self._default_work_thread_pool.shutdown()
     # get_responses may be blocked on responses.get(), but we need to return
     # control to its caller.
-    responses.put(no_more_work)
+    self._responses.put(no_more_work)
     self._data_channel_factory.close()
-    state_handler.done()
+    # Stop all the workers and clean all the associated resources
+    for worker_wrapper in self.worker_wrappers:
+      worker_wrapper.stop()
     logging.info('Done consuming work.')
 
+  @staticmethod
+  def _schedule(thread_pool, work_request, worker, callback):
+    # Need this wrapper to capture the original stack trace.
+    def do_instruction(request):
+      try:
+        return worker.do_instruction(request)
+      except Exception as e:  # pylint: disable=broad-except
+        traceback_str = traceback.format_exc(e)
+        raise Exception('Error processing request. Original traceback '
+                        'is\n%s\n' % traceback_str)
+
+    thread_pool.submit(do_instruction, work_request).add_done_callback(
+        functools.partial(callback, work_request))
+
+  def _handle_response(self, request, response_future):
+    try:
+      response = response_future.result()
+    except Exception as e:  # pylint: disable=broad-except
+      logging.error(
+          'Error processing instruction %s',
+          request.instruction_id,
+          exc_info=True)
+      response = beam_fn_api_pb2.InstructionResponse(
+          instruction_id=request.instruction_id, error=str(e))
+    self._responses.put(response)
+
+  def request_register(self, request):
+    for process_bundle_descriptor in getattr(
+        request, request.WhichOneof('request')).process_bundle_descriptor:
+      self._fns[process_bundle_descriptor.id] = process_bundle_descriptor
+
+    # Pass the response as future to _handle_response.
+    future = futures.Future()
+    future.set_result(
+        beam_fn_api_pb2.InstructionResponse(
+            instruction_id=request.instruction_id,
+            register=beam_fn_api_pb2.RegisterResponse()))
+    self._handle_response(request, future)
+
+  def request_process_bundle(self, request):
+
+    def _next_worker_id():
+      # Get the next worker in round robin fashion
+      self._worker_index = self._worker_index + 1
```

Comment:
<!--new_thread; commit:5dfb3c0f3b174719be925a98b29ef1648d74545c; resolved:0-->
Is it possible to pick a free worker, instead of submitting to the queue of the next worker. It is possible that next worker has existing work but some other worker might be free.